### PR TITLE
[JSON-API] Concurrent query etc. metrics

### LINKED ISSUE #10021: [JSON-API] Add metrics for concurrent requests
From #10020:

> Concurrent requests: This should probably be split into 3 or maybe 4 metrics: concurrent command submissions, concurrent queries, concurrent websocket connections and maybe one for party & package allocation.

### PR DESCRIPTION
**Pull request by realvictorprm**
Should fix #10021

It was fun to play around with the metrics and trying to do some fun stress testing to ensure that the metrics work correctly :smile:

changelog_begin

- [JSON-API] The metrics which describe the amount of these concurrent events is now available:
	- running http request
	- command submissions
	- package allocations
	- party allocations

changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [x] If necessary also did manual testing

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by stefanobaghino-da — APPROVED** [review]

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 1]
```diff
@@ -707,7 +707,11 @@ final class Metrics(val registry: MetricRegistry) {
       private val Prefix: MetricName = daml.Prefix :+ "http_json_api"
 
       val httpRequest: Timer = registry.timer(Prefix :+ "http_request")
+      val httpRequestRunning: Meter = registry.meter(Prefix :+ "http_request_running")
       val websocketRequest: Counter = registry.counter(Prefix :+ "websocket_request")
+      val commandSubmission: Meter = registry.meter(Prefix :+ "command_submission")
+      val packageAllocation: Meter = registry.meter(Prefix :+ "package_allocation")
+      val partyAllocation: Meter = registry.meter(Prefix :+ "party_allocation")
```
Comment (cocreature) [inline ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:714]: this seems somewhat inconsistent. For requests you suffix the concurrent counter with `_running`. For everything else you don’t. For websockets, `_request` is a counter for http it’s a timer.
I think what would help is some consistency and maybe some comments here on what those metrics represent.

Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:714]: you're right, I oversaw that one. I'll change that.

Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala]: addressed with ce07e09

**Comment by realvictorprm** [comment]
thanks for your input on the metric names @cocreature, as I went over them again I decided to rename the running ones to throughput as this accurately reflects what the "Meter" actually meters :smile:

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala (hunk 1 above)
Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala]: I could also rename the timers with a suffix `_timing`.

Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala]: I think I'm doing that as I prefer better names right now.

Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala]: ok done with 0a40a8b

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 2]
```diff
@@ -706,8 +706,19 @@ final class Metrics(val registry: MetricRegistry) {
     object HttpJsonApi {
       private val Prefix: MetricName = daml.Prefix :+ "http_json_api"
 
-      val httpRequest: Timer = registry.timer(Prefix :+ "http_request")
-      val websocketRequest: Counter = registry.counter(Prefix :+ "websocket_request")
+      // Meters how long processing of a request takes
+      val httpRequestTimer: Timer = registry.timer(Prefix :+ "http_request_timing")
+      // Meters http requests throughput
```
Comment (cocreature) [inline ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:711]: totally fine to do this in a separate PR but I’d probably just replace this by a meter for queries. I don’t think seeing requests of different types grouped together is very useful.

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 3]
```diff
@@ -706,8 +706,19 @@ final class Metrics(val registry: MetricRegistry) {
...
+      val httpRequestThroughput: Meter = registry.meter(Prefix :+ "http_request_throughput")
+      // Meters how many websocket connections are currently active
+      val websocketRequestCounter: Counter = registry.counter(Prefix :+ "websocket_request_count")
+      // Meters command submissions throughput
+      val commandSubmissionThroughput: Meter =
+        registry.meter(Prefix :+ "command_submission_throughput")
+      // Meters package uploads throughput
+      val uploadPackagesThroughput: Meter = registry.meter(Prefix :+ "upload_packages_throughput")
+      // Meters party allocation throughput
+      val allocatePartyThroughput: Meter = registry.meter(Prefix :+ "allocation_party_throughput")
```
Comment (cocreature) [inline ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:721]: For all of those, I think we also want counters but no need to do this in this PR.

**Review by cocreature — APPROVED** [review]
:shipit:

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala (hunk 3 above)
Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:721]: This is planned already, but for the next PR. Want to keep things small so I don't mess up :smile:

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala (hunk 2 above)
Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:711]: I probably adjust this in the next PR where I want to add the timers for the queries.
